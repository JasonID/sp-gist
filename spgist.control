comment = 'SP-GIST index (access method)'
default_version = '1.0'
module_pathname = '$libdir/spgist'
relocatable = true

