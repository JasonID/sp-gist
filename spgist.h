#ifndef _SPGIST_H_
#define _SPGIST_H_

#include "access/genam.h"
#include "access/itup.h"
#include "access/xlog.h"
#include "fmgr.h"


#define SPGIST_PROP_PROC		1
#define SPGIST_CHOOSE_PROC		2
#define SPGIST_PICKSPLIT_PROC	3
#define SPGIST_LEAFCONS_PROC	4
#define SPGIST_INNERCONS_PROC	5
#define SPGISTNProc         	5

typedef struct SpGistOpClassProp 
{
	Oid				leafType;
	Oid				prefixType;
	Oid				nodeType;
} SpGistOpClassProp;

typedef struct SpGistPageOpaqueData
{
	uint16         nParents;
	uint16         flags;
} SpGistPageOpaqueData;

typedef SpGistPageOpaqueData *SpGistPageOpaque;

#define SPGIST_META     	(1<<0)
#define SPGIST_DELETED  	(1<<1)
#define SPGIST_LEAF   		(1<<2)
#define SPGIST_HAS_DELETED  (1<<3)

#define SpGistPageGetOpaque(page) ( (SpGistPageOpaque) PageGetSpecialPointer(page) )
#define SpGistPageGetMaxOffset(page) PageGetMaxOffsetNumber(page)
#define SpGistPageIsMeta(page) ( SpGistPageGetOpaque(page)->flags & SPGIST_META)
#define SpGistPageIsDeleted(page) ( SpGistPageGetOpaque(page)->flags & SPGIST_DELETED)
#define SpGistPageSetDeleted(page)    ( SpGistPageGetOpaque(page)->flags |= SPGIST_DELETED)
#define SpGistPageSetNonDeleted(page) ( SpGistPageGetOpaque(page)->flags &= ~SPGIST_DELETED)
#define SpGistPageIsLeaf(page) ( SpGistPageGetOpaque(page)->flags & SPGIST_LEAF)
#define SpGistPageSetLeaf(page)    ( SpGistPageGetOpaque(page)->flags |= SPGIST_LEAF)
#define SpGistPageSetInner(page) ( SpGistPageGetOpaque(page)->flags &= ~SPGIST_LEAF)
#define SpGistPageGetData(page)      (  (SpGistLeafTuple*)PageGetContents(page) )

#define SPGIST_METAPAGE_BLKNO    (0)
#define SPGIST_HEAD_BLKNO        (1)

typedef BlockNumber FreeBlockNumberArray[
			MAXALIGN_DOWN(
				BLCKSZ -
					SizeOfPageHeaderData -
					MAXALIGN(sizeof(SpGistPageOpaqueData)) -
					/* header of SpGistMetaPageData struct */
					MAXALIGN(sizeof(uint16) * 2 + sizeof(uint32))
			) / sizeof(BlockNumber)
	];

typedef struct SpGistMetaPageData
{
	uint32                  magickNumber;
	uint16                  nStart;
	uint16                  nEnd;
	FreeBlockNumberArray    notFullPage;
} SpGistMetaPageData;

#define SPGIST_MAGICK_NUMBER (0xBA0BABED)

#define SpGistMetaBlockN     (sizeof(FreeBlockNumberArray) / sizeof(BlockNumber))
#define SpGistPageGetMeta(p) \
    ((SpGistMetaPageData *) PageGetContents(p))

typedef struct SpGistTypeDesc 
{
	Oid		type;
	bool	attbyval;
	int16	attlen;
} SpGistTypeDesc;

typedef struct SpGistState
{
	SpGistOpClassProp	prop;
	SpGistTypeDesc		attType;
	SpGistTypeDesc		attNodeType;
	SpGistTypeDesc		attPrefixType; /* optional */
	
	FmgrInfo			chooseFn;
	FmgrInfo			picksplitFn;
	FmgrInfo			leafConsistentFn;
	FmgrInfo			innerConsistentFn;

	TupleDesc			nodeTupDesc;
} SpGistState;

typedef struct SpGistScanOpaqueData
{
	SpGistState  	state;
	MemoryContext 	tempCxt;
} SpGistScanOpaqueData;

typedef SpGistScanOpaqueData *SpGistScanOpaque;

/*
 * node/tuple types
 * Inner tuple layout:
 * (bit of presence of prefix)(n nodes)(size)[(prefix)](array of nodes)
 */
typedef struct SpGistInnerTupleData
{
	unsigned int	hasPrefix:1, /* could be true for TreeShrink */ 
					nNodes:15,
					size:16;
	char			data[1]; /* variable size */
} SpGistInnerTupleData;
typedef SpGistInnerTupleData *SpGistInnerTuple;

#define SGITHDRSZ			MAXALIGN(offsetof(SpGistInnerTupleData, data))
#define  _SGITDATA(x)		( ((char*)(x)) + SGITHDRSZ )
#define SGITDATAPTR(x)  	( ((x)->hasPrefix) ? _SGITDATA(x) : NULL )
#define SGITDATUM(x, s)		( ((x)->hasPrefix) ? \
								( \
									((s)->attPrefixType.attbyval) ? \
									( *(Datum*)_SGITDATA(x) ) \
									: \
									PointerGetDatum(_SGITDATA(x)) \
								) \
								: (Datum)0 )
#define SGITNODEPTR(x, s) 	( (IndexTuple)( \
								_SGITDATA(x) + \
								( ((x)->hasPrefix) ? \
									getTypeLength( &(s)->attPrefixType, PointerGetDatum(_SGITDATA(x)) ) \
									: 0 \
								) \
							) )

#define SGITITERATE(x, s, i, it)	\
	for((i) = 0, (it)= SGITNODEPTR((x),(s)) ; \
		(i) < (x)->nNodes; \
		(i)++, (it) = (IndexTuple)(((char*)(it)) + IndexTupleSize(it))) 


/*
 * indexed value could be empty (not a NULL) if it's
 * fully encoded in a tree path
 */
typedef struct SpGistLeafTupleData
{
	ItemPointerData	heapPtr;
	OffsetNumber	nextOffset;
	unsigned short 	size;
	char			data[1]; /* variable size */
} SpGistLeafTupleData;
typedef SpGistLeafTupleData *SpGistLeafTuple;

#define SGLTHDRSZ			MAXALIGN(offsetof(SpGistLeafTupleData, data))
#define SGLTDATAPTR(x)  	( ((char*)(x)) + SGLTHDRSZ )
#define SGLTDATUM(x, s)		( ((s)->attType.attbyval) ? \
								( *(Datum*)SGLTDATAPTR(x) ) \
								: \
								PointerGetDatum(SGLTDATAPTR(x)) \
							)

/*
 * interface struct
 */

typedef struct spgChooseIn
{
	Datum	datum; /* full indexing datum */
	int		level;

	/*
	 * Inner tuple data
	 */
	bool	hasPrefix;
	Datum	prefixDatum;

	int		nNodes;
	Datum	*nodeDatums;
} spgChooseIn;

typedef enum spgChooseResultType
{
	spgMatchNode = 0,
	spgAddNode,
	spgSplitTuple
} spgChooseResultType;

typedef struct spgChooseOut
{
	spgChooseResultType	resultType;
	union {
		struct {
			int		nodeN;
			int		levelAdd;
			Datum	restDatum;
		} matchNode;
		struct {
			Datum	nodeDatum;
		} addNode;
		struct {
			/* new inner tuple with one node */
			int		levelPrefixAdd;
			bool	prefixHasPrefix;
			Datum	prefixPrefixDatum;
			Datum	nodeDatum;

			/* new next inner tuple with all old nodes */ 
			int		levelPostfixAdd;
			bool	postfixHasPrefix;
			Datum	postfixPrefixDatum;
		} splitTuple;
	} result; 
} spgChooseOut;

typedef struct spgPickSplitIn
{
	int		nTuples;
	Datum	*datums;
} spgPickSplitIn;

typedef struct spgPickSplitOut
{
	bool	hasPrefix;
	Datum	prefixDatum;

	int		nNodes;
	Datum	*nodeDatums;

	int		*mapTuplesToNodes;
	Datum	*leafTupleDatums;
} spgPickSplitOut;

typedef struct spgInnerConsistentIn
{
	Datum	query;

	int		level;

	bool	hasPrefix;
	Datum	prefixDatum;

	int		nNodes;
	Datum	*nodeDatums;
} spgInnerConsistentIn;

typedef struct spgInnerConsistentOut
{
	int		nNodes;
	int		levelAdd; /* including node's levels */
	int		*nodeNumbers;
} spgInnerConsistentOut;

/* spgutils.h */
void initSpGistState(SpGistState *state, Relation index);
Buffer SpGistNewBuffer(Relation index);
void SpGistInitBuffer(Buffer b, uint16 f);
void SpGistInitPage(Page page, uint16 f, Size pageSize);
void SpGistInitMetabuffer(Buffer b, Relation index);

unsigned int getTypeLength(SpGistTypeDesc *att, Datum datum);
SpGistLeafTuple spgFormLeafTuple(SpGistState *state, ItemPointer heapPtr, Datum datum);
SpGistInnerTuple spgFormInnerTuple(SpGistState *state, bool hasPrefix, Datum prefix, 
									int nNodes, IndexTuple *nodes);

void spgdoinsert(Relation index, SpGistState *state, ItemPointer heapPtr, Datum datum);
#endif
