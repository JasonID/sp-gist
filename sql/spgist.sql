CREATE EXTENSION spgist;

CREATE TABLE test_text(t text);

\copy test_text from 'data/text.data'

CREATE INDEX ttidx ON test_text USING spgist (t);

SET enable_seqscan=off;

EXPLAIN (COSTS OFF)
SELECT * FROM test_text WHERE t = 'http://0-2000webhosting.co.uk/email-configuration.htm';

SELECT * FROM test_text WHERE t = 'http://0-2000webhosting.co.uk/email-configuration.htm';

SELECT * FROM test_text WHERE t = 'http://www.data-wales.co.uk/lamb.htm';

SELECT * FROM test_text WHERE t = 'http://comwebhosting.co.uk/plan-order.asp?plan=Windows_E_Biz';

SELECT * FROM test_text WHERE t = 'http://abcde.co.uk/betterhearingservices/about_us.html';

SELECT * FROM test_text WHERE t = 'http://www.airportparkingexpress.co.uk/belfast.htm';

SELECT * FROM test_text WHERE t = 'http://www.skiing-holidays-austria.co.uk/16/panorama-ski-austria-from-prestwick.html';

SELECT * FROM test_text WHERE t = 'http://www.all-inclusive-holiday-bargains.co.uk/all-inclusive-holidays-austria.htm';


CREATE TABLE test_quad(p point);

\copy test_quad from 'data/point.data'

CREATE INDEX tqidx ON test_quad USING spgist (p);

EXPLAIN (COSTS OFF)
SELECT * FROM test_quad WHERE p ~= '(8.51277472174491,5.86434731598175)';

SELECT * FROM test_quad WHERE p ~= '(8.51277472174491,5.86434731598175)';

SELECT * FROM test_quad WHERE p ~= '(1.39955907884019,9.12045046572942)';
