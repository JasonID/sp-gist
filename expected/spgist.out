CREATE EXTENSION spgist;
CREATE TABLE test_text(t text);
\copy test_text from 'data/text.data'
CREATE INDEX ttidx ON test_text USING spgist (t);
SET enable_seqscan=off;
EXPLAIN (COSTS OFF)
SELECT * FROM test_text WHERE t = 'http://0-2000webhosting.co.uk/email-configuration.htm';
                                       QUERY PLAN                                        
-----------------------------------------------------------------------------------------
 Bitmap Heap Scan on test_text
   Recheck Cond: (t = 'http://0-2000webhosting.co.uk/email-configuration.htm'::text)
   ->  Bitmap Index Scan on ttidx
         Index Cond: (t = 'http://0-2000webhosting.co.uk/email-configuration.htm'::text)
(4 rows)

SELECT * FROM test_text WHERE t = 'http://0-2000webhosting.co.uk/email-configuration.htm';
                           t                           
-------------------------------------------------------
 http://0-2000webhosting.co.uk/email-configuration.htm
(1 row)

SELECT * FROM test_text WHERE t = 'http://www.data-wales.co.uk/lamb.htm';
                  t                   
--------------------------------------
 http://www.data-wales.co.uk/lamb.htm
(1 row)

SELECT * FROM test_text WHERE t = 'http://comwebhosting.co.uk/plan-order.asp?plan=Windows_E_Biz';
                              t                               
--------------------------------------------------------------
 http://comwebhosting.co.uk/plan-order.asp?plan=Windows_E_Biz
(1 row)

SELECT * FROM test_text WHERE t = 'http://abcde.co.uk/betterhearingservices/about_us.html';
                           t                            
--------------------------------------------------------
 http://abcde.co.uk/betterhearingservices/about_us.html
(1 row)

SELECT * FROM test_text WHERE t = 'http://www.airportparkingexpress.co.uk/belfast.htm';
                         t                          
----------------------------------------------------
 http://www.airportparkingexpress.co.uk/belfast.htm
(1 row)

SELECT * FROM test_text WHERE t = 'http://www.skiing-holidays-austria.co.uk/16/panorama-ski-austria-from-prestwick.html';
                                          t                                           
--------------------------------------------------------------------------------------
 http://www.skiing-holidays-austria.co.uk/16/panorama-ski-austria-from-prestwick.html
(1 row)

SELECT * FROM test_text WHERE t = 'http://www.all-inclusive-holiday-bargains.co.uk/all-inclusive-holidays-austria.htm';
                                         t                                          
------------------------------------------------------------------------------------
 http://www.all-inclusive-holiday-bargains.co.uk/all-inclusive-holidays-austria.htm
(1 row)

CREATE TABLE test_quad(p point);
\copy test_quad from 'data/point.data'
CREATE INDEX tqidx ON test_quad USING spgist (p);
EXPLAIN (COSTS OFF)
SELECT * FROM test_quad WHERE p ~= '(8.51277472174491,5.86434731598175)';
                               QUERY PLAN                                
-------------------------------------------------------------------------
 Bitmap Heap Scan on test_quad
   Recheck Cond: (p ~= '(8.51277472174491,5.86434731598175)'::point)
   ->  Bitmap Index Scan on tqidx
         Index Cond: (p ~= '(8.51277472174491,5.86434731598175)'::point)
(4 rows)

SELECT * FROM test_quad WHERE p ~= '(8.51277472174491,5.86434731598175)';
                  p                  
-------------------------------------
 (8.51277472174491,5.86434731598175)
(1 row)

SELECT * FROM test_quad WHERE p ~= '(1.39955907884019,9.12045046572942)';
                  p                  
-------------------------------------
 (1.39955907884019,9.12045046572942)
(1 row)

